rust/catalyst-types/src/id_uri/role_index.rs

```diff
@@ -12,54 +12,135 @@ use thiserror::Error;
 #[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
-pub struct RoleIndex(u16);
+#[repr(u8)]
+#[non_exhaustive]
+pub enum RoleId {
+    /// Primary required role use for voting and commenting.
+    Role0 = 0,
+
+    /// Delegated representative (dRep) that vote on behalf of delegators.
+    DelegatedRepresentative = 1,
 
-impl RoleIndex {
-    /// Delegated Representative
-    pub const DREP: RoleIndex = RoleIndex(1);
-    /// Proposer
-    pub const PROPOSER: RoleIndex = RoleIndex(3);
-    /// Default Role 0
-    pub const ROLE_0: RoleIndex = RoleIndex(0);
+    /// Proposer that enabling creation, collaboration, and submission of proposals.
+    Proposer = 3,
+
+    /// A custom role.
+    Custom(u8),
+}
 
-    /// Is the `RoleIndex` the default value (Role 0)
+impl RoleId {
+    /// Is the `RoleId` the default value [`Self::Role0`]
     #[must_use]
     pub fn is_default(self) -> bool {
-        self == Self::ROLE_0
+        self == Self::Role0
+    }
+
+    /// Returns the `u8` representation of the role.
+    #[must_use]
+    pub const fn as_u8(self) -> u8 {
+        match self {
+            RoleId::Role0 => 0,
+            RoleId::DelegatedRepresentative => 1,
+            RoleId::Proposer => 3,
+            RoleId::Custom(b) => b,
+        }
+    }
+
+    /// Returns `true` if the role belongs to the canonical set of known roles.
+    #[must_use]
+    pub const fn is_known(self) -> bool {
+        matches!(
+            self,
+            Self::Role0 | Self::DelegatedRepresentative | Self::Proposer
+        )
     }
 }
 
-impl Default for RoleIndex {
+impl Default for RoleId {
     fn default() -> Self {
-        Self::ROLE_0
+        Self::Role0
     }
 }
 
-impl From<u16> for RoleIndex {
-    fn from(value: u16) -> Self {
-        Self(value)
+impl From<u8> for RoleId {
+    fn from(value: u8) -> Self {
+        match value {
+            0 => Self::Role0,
+            1 => Self::DelegatedRepresentative,
+            3 => Self::Proposer,
+            b => Self::Custom(b),
+        }
     }
 }
 
-impl FromStr for RoleIndex {
-    type Err = RoleIndexError;
+impl FromStr for RoleId {
+    type Err = RoleIdError;
 
     fn from_str(s: &str) -> Result<Self, Self::Err> {
-        Ok(Self(s.parse::<u16>()?))
+        let value: u8 = s.parse()?;
+
+        Ok(Self::from(value))
     }
 }
 
-impl Display for RoleIndex {
+impl Display for RoleId {
     fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), std::fmt::Error> {
-        write!(f, "{}", self.0)
+        write!(f, "{}", self.as_u8())
+    }
+}
+
+impl<'a, C> minicbor::Decode<'a, C> for RoleId {
+    fn decode(d: &mut minicbor::Decoder<'a>, ctx: &mut C) -> Result<Self, minicbor::decode::Error> {
+        <u8 as minicbor::Decode<'a, C>>::decode(d, ctx).map(Self::from)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use minicbor::{Decoder, Encoder};
+    use proptest::{num, prelude::*};
+
+    use super::*;
+
+    proptest! {
```

Comment:
Use `proptest` attribute macro like this
https://github.com/proptest-rs/proptest/blob/4672257f7b01ef3e4b7df16359bb7c9b0d9685f5/proptest/tests/pass/simple_example.rs#L4